### PR TITLE
Improve exception throwing for System.Collections

### PR DESCRIPTION
**Pull request by hughbe**
- Merge identical exception messages to remove string resources
- Fix incorrect exception parameter names
- Fix incorrect exception messages
- Add missing parameter names to several exceptions
- Add object to ArgumentOutOfRangeExceptions and the like

### CODE REVIEW

**Comment by hughbe** [comment]
Correct me if I'm wrong, but I just realised this may fix #4940

**Comment by JonHanna** [comment]
I think so, but I'm not convinced that's a good thing. If indices are calculated as offsets then exceptions will contain an incorrect index. Consider:

```C#
public T this[int index]
{
    get
    {
        int len = _first.Count;
        return index < len ? _first[index] : _second[index - len];
    }
}
```

Right now that would be correct code that either works correctly or gives a reasonable exception. If `_second` is a type that includes indices in out of range exceptions then e.g if `_first` and `_second` both have 10 items, making the combined list 20 in size and code looks for the item with index `23` the exception will say the incorrect index was `3` which should have been well within the range [0, 19] and so cause more confusion than including an index could resolve.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 1]
```diff
@@ -249,7 +249,7 @@ private void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
 
             if (index < 0 || index > array.Length)
             {
-                throw new ArgumentOutOfRangeException("index", SR.ArgumentOutOfRange_NeedNonNegNum);
+                throw new ArgumentOutOfRangeException("index", index, SR.ArgumentOutOfRange_NeedNonNegNum);
```
Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: This choice of message is a bug in the original (also present in full framework) in that `((ICollection)new Dictionary<string, int>()).CopyTo(new KeyValuePair<string, int>[3], 100);` complains "Non-negative number required.". If you're going to change it, may as well fix it.

Comment (hughbe) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Whoops - I found and changed several exception messages like this but this one slipped past my radar. I'll change it soon

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs [hunk 2]
```diff
@@ -188,7 +188,7 @@ public void CopyTo(T[] array, int index)
 
             if (index < 0 || index > array.Length)
```
Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: There's a micro-opt possible with this patter where one uses `if ((uint)index > (uint)array.Length)` or equivalent. While at the C# level that's got two casts, it compiles to a `cgt.un` rather than any explicit caching operation, so the effect is one less branch and one less comparison in the most common case.
Whether that's worth doing or not is another question.

**Comment by JonHanna** [comment]
Test  Innerloop OSX Release Build and Test
Test Innerloop Windows_NT Debug Build and Test
Test Innerloop Windows_NT Release Build and Test

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: So this works because if int is <0, the uint cast is a huge overflowed number right?
Does the performance benefit outweigh the slight lack of readability?

Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: That would be a subjective question, with a subjective answer. I don't find it hurts readability, but then I'm used to it, so I'm probably a bad data-point! Other people might disagree strongly.

**Comment by JonHanna** [comment]
Aside from keeping the erroneous message in Dictionary.cs, LGTM.

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs (hunk 2 above)
Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: (But yes, it overflows to something in the range `int.MaxValue + 1` to `uint.MaxValue` so as long as the other comparand is guaranteed to be in the range `0` to `int.MaxValue` it gives the same result).

Comment (ianhays) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: IMO it does detract from the readability but not insofar that it becomes confusing.

**Comment by ianhays** [comment]
LGTM.

**Comment by hughbe** [comment]
Any progress on this?

**Comment by ianhays** [comment]
@stephentoub

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 3]
```diff
@@ -247,9 +247,9 @@ private void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
                 throw new ArgumentNullException("array");
             }
 
-            if (index < 0 || index > array.Length)
+            if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Why this change?  To me the behavior isn't as clear, and I doubt it makes any measurable perf difference.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 4]
```diff
@@ -494,17 +494,17 @@ void ICollection.CopyTo(Array array, int index)
...
             }
 
             if (array.GetLowerBound(0) != 0)
             {
-                throw new ArgumentException(SR.Arg_NonZeroLowerBound);
+                throw new ArgumentException(SR.Arg_NonZeroLowerBound, "array");
             }
 
-            if (index < 0 || index > array.Length)
+            if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 5]
```diff
@@ -855,9 +855,9 @@ public void CopyTo(TKey[] array, int index)
                     throw new ArgumentNullException("array");
                 }
 
-                if (index < 0 || index > array.Length)
+                if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Ditto.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 6]
```diff
@@ -923,17 +923,17 @@ void ICollection.CopyTo(Array array, int index)
...
                 }
 
                 if (array.GetLowerBound(0) != 0)
                 {
-                    throw new ArgumentException(SR.Arg_NonZeroLowerBound);
+                    throw new ArgumentException(SR.Arg_NonZeroLowerBound, "array");
                 }
 
-                if (index < 0 || index > array.Length)
+                if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Ditto.

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 7]
```diff
@@ -1084,9 +1084,9 @@ public void CopyTo(TValue[] array, int index)
                     throw new ArgumentNullException("array");
                 }
 
-                if (index < 0 || index > array.Length)
+                if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs [hunk 8]
```diff
@@ -1152,17 +1152,17 @@ void ICollection.CopyTo(Array array, int index)
...
                 }
 
                 if (array.GetLowerBound(0) != 0)
                 {
-                    throw new ArgumentException(SR.Arg_NonZeroLowerBound);
+                    throw new ArgumentException(SR.Arg_NonZeroLowerBound, "array");
                 }
 
-                if (index < 0 || index > array.Length)
+                if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs [hunk 9]
```diff
@@ -186,9 +186,9 @@ public void CopyTo(T[] array, int index)
                 throw new ArgumentNullException("array");
             }
 
-            if (index < 0 || index > array.Length)
+            if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/List.cs [hunk 10]
```diff
@@ -567,22 +567,19 @@ public int FindLastIndex(int startIndex, int count, Predicate<T> match)
                 // Special case for 0 length List
                 if (startIndex != -1)
                 {
-                    throw new ArgumentOutOfRangeException("startIndex", SR.ArgumentOutOfRange_Index);
+                    throw new ArgumentOutOfRangeException("startIndex", startIndex, SR.ArgumentOutOfRange_Index);
                 }
             }
-            else
+            // Make sure we're not out of range
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Nit: we try not to have comments in a position like this.  It'd be fine at the end of the ```else if``` line

File: src/System.Collections/src/System/Collections/Generic/List.cs [hunk 11]
```diff
@@ -1093,12 +1090,12 @@ public void Sort(int index, int count, IComparer<T> comparer)
...
             if (index < 0)
             {
-                throw new ArgumentOutOfRangeException("index", SR.ArgumentOutOfRange_NeedNonNegNum);
+                throw new ArgumentOutOfRangeException("index", index, SR.ArgumentOutOfRange_NeedNonNegNum);
             }
 
             if (count < 0)
             {
-                throw new ArgumentOutOfRangeException("count", SR.ArgumentOutOfRange_NeedNonNegNum);
+                throw new ArgumentOutOfRangeException("count", index, SR.ArgumentOutOfRange_NeedNonNegNum);
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Why is this ```index``` rather than ```count```?

File: src/System.Collections/src/System/Collections/Generic/List.cs [hunk 12]
```diff
@@ -1037,12 +1034,12 @@ public void Reverse(int index, int count)
...
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: Why is this ```index``` rather than ```count```?

File: src/System.Collections/src/System/Collections/Generic/Queue.cs [hunk 13]
```diff
@@ -154,18 +154,18 @@ void System.Collections.ICollection.CopyTo(Array array, int index)
...
 
             if (array.GetLowerBound(0) != 0)
             {
-                throw new ArgumentException(SR.Arg_NonZeroLowerBound);
+                throw new ArgumentException(SR.Arg_NonZeroLowerBound, "array");
             }
 
             int arrayLen = array.Length;
-            if (index < 0 || index > arrayLen)
+            if ((uint)index > (uint)arrayLen)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Queue.cs]: Ditto on why we're changing these.... I realize it's done like this elsewhere in the code, but I still find the original more easily understood.

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 14]
```diff
@@ -443,19 +443,19 @@ public bool ContainsValue(TValue value)
         }
 
         // Copies the values in this SortedList to an array.
-        void ICollection<KeyValuePair<TKey, TValue>>.CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex)
+        void ICollection<KeyValuePair<TKey, TValue>>.CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Why are you changing the parameter name?  It's named ```arrayIndex``` in the interface itself.  Unless there's a really good reason to change it, we should leave it as is.

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 15]
```diff
@@ -443,19 +443,19 @@ public bool ContainsValue(TValue value)
...
         {
             if (array == null)
             {
                 throw new ArgumentNullException("array");
             }
 
-            if (arrayIndex < 0 || arrayIndex > array.Length)
+            if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Ditto on the comparison change.

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 16]
```diff
@@ -476,20 +476,20 @@ void System.Collections.ICollection.CopyTo(Array array, int arrayIndex)
...
             }
 
             if (array.GetLowerBound(0) != 0)
             {
-                throw new ArgumentException(SR.Arg_NonZeroLowerBound);
+                throw new ArgumentException(SR.Arg_NonZeroLowerBound, "array");
             }
 
-            if (arrayIndex < 0 || arrayIndex > array.Length)
+            if ((uint)index > (uint)array.Length)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 17]
```diff
@@ -544,8 +544,8 @@ private void EnsureCapacity(int min)
         // 
         private TValue GetByIndex(int index)
         {
-            if (index < 0 || index >= _size)
-                throw new ArgumentOutOfRangeException("index", SR.ArgumentOutOfRange_Index);
+            if ((uint)index >= (uint)_size)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 18]
```diff
@@ -575,7 +575,8 @@ System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
         // 
         private TKey GetKey(int index)
         {
-            if (index < 0 || index >= _size) throw new ArgumentOutOfRangeException("index", SR.ArgumentOutOfRange_Index);
+            if ((uint)index >= (uint)_size)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 19]
```diff
@@ -711,7 +712,8 @@ public bool TryGetValue(TKey key, out TValue value)
         // 
         public void RemoveAt(int index)
         {
-            if (index < 0 || index >= _size) throw new ArgumentOutOfRangeException("index", SR.ArgumentOutOfRange_Index);
+            if ((uint)index >= (uint)_size)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Ditto

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs (hunk 3 above)
Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: I suggested it as a possibility at https://github.com/dotnet/corefx/pull/5672#discussion_r51124178
If someone finds it unclear I think it shouldn't be done. (Its become an idiom to me, so I can't really judge how clear or unclear it is).

Comment (JonHanna) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: I suggested it as a possibility at https://github.com/dotnet/corefx/pull/5672#discussion_r51124178
If someone finds it unclear I think it shouldn't be done. (Its become an idiom to me, so I can't really judge how clear or unclear it is).

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 20]
```diff
@@ -675,17 +675,17 @@ void ICollection.CopyTo(Array array, int index)
...
             if (array.Rank != 1)
             {
-                throw new ArgumentException(SR.Arg_RankMultiDimNotSupported);
+                throw new ArgumentException(SR.Arg_RankMultiDimNotSupported, "array");
             }
 
             if (array.GetLowerBound(0) != 0)
             {
-                throw new ArgumentException(SR.Arg_NonZeroLowerBound);
+                throw new ArgumentException(SR.Arg_NonZeroLowerBound, "index");
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Why ```index``` rather than ```array```?

File: src/System.Collections/src/System/Collections/Generic/Dictionary.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/Dictionary.cs]: I get that it means the same thing, but it's very clear to me when I see the other comparisons exactly what's being checked for, whereas this takes a bit longer to think it through.  I'd prefer the original. (I'm not, however, suggesting that all of the existing occurrences of this pattern be changed.)

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 21]
```diff
@@ -1891,7 +1891,7 @@ public virtual SortedSet<T> GetViewBetween(T lowerValue, T upperValue)
         {
             if (Comparer.Compare(lowerValue, upperValue) > 0)
             {
-                throw new ArgumentException("lowerBound is greater than upperBound");
+                throw new ArgumentException("lowerValue is greater than upperValue", "lowerValue");
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: As long as you're changing something here, this string should be in a resource.

File: src/System.Collections/src/System/Collections/Generic/List.cs [hunk 22]
```diff
@@ -740,7 +737,7 @@ public void Insert(int index, T item)
             // Note that insertions at the end are legal.
             if ((uint)index > (uint)_size)
             {
-                throw new ArgumentOutOfRangeException("index", SR.ArgumentOutOfRange_ListInsert);
+                throw new ArgumentOutOfRangeException("index", index, SR.ArgumentOutOfRange_Index);
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/List.cs]: This doesn't look like a valid change.  ArgumentOutOfRange_Index talks about the index being less than the size of the collection, but here it can be equal to the size of the collection.  I'd leave this one as it was.

File: src/System.Collections/src/System/Collections/Generic/LinkedList.cs [hunk 23]
```diff
@@ -186,9 +186,9 @@ public void CopyTo(T[] array, int index)
...
             {
-                throw new ArgumentOutOfRangeException("index", SR.Format(SR.IndexOutOfRange, index));
+                throw new ArgumentOutOfRangeException("index", index, SR.ArgumentOutOfRange_Index);
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/LinkedList.cs]: This doesn't look valid, either.  ArgumentOutOfRange_Index says the index must be less than the size, but it's valid here for it to be equal to the size.

**Comment by hughbe** [comment]
Before I go back and undo all the changes Jonnsuggested about condensing the out of range checks, are both of you two reviewers sure?

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs (hunk 14 above)
Comment (hughbe) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Whoops, I was looking at the non-generic ICollection interface. I'll undo this

**Comment by hughbe** [comment]
Responded to PR feedback and rebased

File: src/System.Collections/src/Resources/Strings.resx [hunk 24]
```diff
@@ -202,7 +193,7 @@
     <value>Target array type is not compatible with the type of items in the collection.</value>
   </data>
   <data name="ArgumentOutOfRange_BiggerThanCollection" xml:space="preserve">
-    <value>Larger than collection size.</value>
+    <value>Must be less than or equal to the size of the collection.</value>
```
Comment (stephentoub) [inline src/System.Collections/src/Resources/Strings.resx]: Neither the old nor the new wording seems to make sense here:
https://github.com/dotnet/corefx/pull/5672/files#diff-78d6503f5065af177a325d9668fe4735L903
or here:
https://github.com/dotnet/corefx/pull/5672/files#diff-78d6503f5065af177a325d9668fe4735R904

File: src/System.Collections/src/Resources/Strings.resx [hunk 25]
```diff
@@ -228,4 +213,7 @@
   <data name="NotSupported_SortedListNestedWrite" xml:space="preserve">
     <value>This operation is not supported on SortedList nested types because they require modifying the original SortedList.</value>
   </data>
+  <data name="SortedSet_LowerValueGreaterThanUpperValue" xml:space="preserve">
+    <value>Must be less than or equal to than upperValue.</value>
```
Comment (stephentoub) [inline src/System.Collections/src/Resources/Strings.resx]: ```to than upperValue``` => ```to upperValue```

File: src/System.Collections/src/Resources/Strings.resx (hunk 24 above)
Comment (hughbe) [inline src/System.Collections/src/Resources/Strings.resx]: I fixed the 1st one, not really sure about the 2nd though. Is "Argument_InvalidOffLen" - "Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection."

Comment (stephentoub) [inline src/System.Collections/src/Resources/Strings.resx]: For the second I'd think ArgumentOutOfRange_Count would be better, with the error mentioning "count" rather than "index".  Argument_InvalidOffLen doesn't work in part because it refers to names that don't exist in this function.

Comment (hughbe) [inline src/System.Collections/src/Resources/Strings.resx]: Done!

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs [hunk 26]
```diff
@@ -453,7 +453,7 @@ public bool ContainsValue(TValue value)
 
             if (arrayIndex < 0 || arrayIndex > array.Length)
             {
-                throw new ArgumentOutOfRangeException("arrayIndex", SR.ArgumentOutOfRange_NeedNonNegNum);
+                throw new ArgumentOutOfRangeException("index", arrayIndex, SR.ArgumentOutOfRange_Index);
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: Why did this change from "arrayIndex" to "index"?

**Comment by stephentoub** [comment]
LGTM other than my one additional comment. Please fix that in a separate commit so I don't need to re-review the whole thing again :wink:

File: src/System.Collections/src/System/Collections/Generic/SortedList.cs (hunk 26 above)
Comment (hughbe) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: That's the suggested name for the ICollection interface. I undid this change and pushed another commit :)

Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: I think you misunderstood me... the ```ICollection<T>.CopyTo``` argument was fine named ```arrayIndex```, but you were then throwing an exception with the string ```"index"``` not ```"arrayIndex"```, so it didn't actually match.

Comment (hughbe) [inline src/System.Collections/src/System/Collections/Generic/SortedList.cs]: oh my bad. fixed

**Comment by stephentoub** [comment]
LGTM. Thanks!

**Comment by stephentoub** [comment]
@dotnet-bot test this please
